### ISSUE TITLE
Vulnerability Scanner - Test comparability with content changes

### ISSUE THREAD

**Issue by sebasfalcone** [wazuh/wazuh#23846]
## Description
We should test the changes from the POC:
- https://github.com/wazuh/intelligence-platform/issues/1573

Doesn't affect the vulnerability scanner capabilities. 

### Testing
To do so, we need to:
1. Ingest the [snapshot generated from the POC](https://github.com/wazuh/wazuh/issues/23846#issuecomment-2145798906) using [4.8.0](https://github.com/wazuh/wazuh/tree/4.8.0)
2. Export the vulnerability list
3. Ingest the [snapshot generated from the POC](https://github.com/wazuh/wazuh/issues/23846#issuecomment-2145798906) using [this branch](https://github.com/wazuh/wazuh/tree/custom/address-nvd-delay-in-cve-updates) (which handles the content changes)
4. Export the vulnerability list
5. Compare both lists
6. PS: Execute the regular content in 4.8.0 and use it as a control

We should arrive at the same result

> [!NOTE]
> As an extra, please check the size of the generated DB in botch cases

## DoD
- [x] Testing performed

**Comment by jftuduri** [comment]
Here is the snapshot: https://drive.google.com/file/d/1cqPRUwTTPR5Gm8XrRJLjuIBF2eA939SP/view?usp=sharing

- 219440 offsets
  - 219383 CVEs
  - 54 translations
  - 3 base rules

**Comment by pereyra-m** [comment]
## Update

I begin with the issue. I'll download the snapshot and perform the requested verifications.

**Comment by pereyra-m** [comment]
## Results

### PoC content / 4.8.0

```
root@fb659c773250:/var/ossec# du -h ./queue/vd/feed/ 
6.3G	./queue/vd/feed/
```

```
root@fb659c773250:/workspaces/wazuh# python3 run.py 
Getting candidates for column: nvd . Result: 342479
Getting candidates for column: suse_server_15 . Result: 260026
Getting candidates for column: suse_desktop_15 . Result: 158671
Getting candidates for column: suse_server_12 . Result: 222974
Getting candidates for column: debian . Result: 1631884
Getting candidates for column: redhat_8 . Result: 266465
Getting candidates for column: redhat_6 . Result: 164003
Getting candidates for column: redhat_7 . Result: 219563
Getting candidates for column: suse_desktop_12 . Result: 46486
Getting candidates for column: canonical . Result: 778312
Getting candidates for column: pypi . Result: 3937
Getting candidates for column: suse_server_11 . Result: 103732
Getting candidates for column: redhat_9 . Result: 236338
Getting candidates for column: alas_1 . Result: 54272
Getting candidates for column: redhat_5 . Result: 25842
Getting candidates for column: suse_desktop_11 . Result: 18832
Getting candidates for column: arch . Result: 8498
Getting candidates for column: rocky . Result: 87358
Getting candidates for column: alma_8 . Result: 76563
Getting candidates for column: alma_9 . Result: 21905
Getting candidates for column: alas_2023 . Result: 22974
Getting candidates for column: alas_2 . Result: 36783
Getting candidates for column: npm . Result: 2673
Getting candidates for column: alas_2022 . Result: 10754
Getting candidates for column: DISCARDED_CNA . Result: 2844
Getting candidates for column: suse . Result: 44
Getting candidates for column: cve5 . Result: 219391
```

###  Regular Content / 4.8.0

```
root@fb659c773250:/var/ossec# du -h ./queue/vd/feed/ 
5.9G	./queue/vd/feed/
```

```
root@fb659c773250:/workspaces/wazuh# python3 run.py 
Getting candidates for column: nvd . Result: 671036
Getting candidates for column: suse_server_15 . Result: 237760
Getting candidates for column: suse_desktop_15 . Result: 151264
Getting candidates for column: suse_server_12 . Result: 204583
Getting candidates for column: debian . Result: 1272145
Getting candidates for column: redhat_8 . Result: 187653
Getting candidates for column: redhat_6 . Result: 142176
Getting candidates for column: redhat_7 . Result: 176155
Getting candidates for column: suse_desktop_12 . Result: 46593
Getting candidates for column: canonical . Result: 592255
Getting candidates for column: pypi . Result: 3533
Getting candidates for column: suse_server_11 . Result: 104336
Getting candidates for column: redhat_9 . Result: 99747
Getting candidates for column: alas_1 . Result: 54495
Getting candidates for column: redhat_5 . Result: 26131
Getting candidates for column: suse_desktop_11 . Result: 18934
Getting candidates for column: arch . Result: 8530
Getting candidates for column: rocky . Result: 85459
Getting candidates for column: alma_8 . Result: 74627
Getting candidates for column: alma_9 . Result: 19337
Getting candidates for column: alas_2023 . Result: 21186
Getting candidates for column: alas_2 . Result: 35584
Getting candidates for column: npm . Result: 2558
Getting candidates for column: alas_2022 . Result: 10826
Getting candidates for column: DISCARDED_CNA . Result: Couldn't find column family: 'DISCARDED_CNA'
0
Getting candidates for column: suse . Result: 13
Getting candidates for column: cve5 . Result: 238605
```

### PoC Content / custom branch

```
root@fb659c773250:/var/ossec# du -h ./queue/vd/feed/ 
6.3G	./queue/vd/feed/
```

```
Getting candidates for column: nvd . Result: 342479
Getting candidates for column: suse_server_15 . Result: 260026
Getting candidates for column: suse_desktop_15 . Result: 158671
Getting candidates for column: suse_server_12 . Result: 222974
Getting candidates for column: debian . Result: 1631884
Getting candidates for column: redhat_8 . Result: 266465
Getting candidates for column: redhat_6 . Result: 164003
Getting candidates for column: redhat_7 . Result: 219563
Getting candidates for column: suse_desktop_12 . Result: 46486
Getting candidates for column: canonical . Result: 778312
Getting candidates for column: pypi . Result: 3937
Getting candidates for column: suse_server_11 . Result: 103732
Getting candidates for column: redhat_9 . Result: 236338
Getting candidates for column: alas_1 . Result: 54272
Getting candidates for column: redhat_5 . Result: 25842
Getting candidates for column: suse_desktop_11 . Result: 18832
Getting candidates for column: arch . Result: 8498
Getting candidates for column: rocky . Result: 87358
Getting candidates for column: alma_8 . Result: 76563
Getting candidates for column: alma_9 . Result: 21905
Getting candidates for column: alas_2023 . Result: 22974
Getting candidates for column: alas_2 . Result: 36783
Getting candidates for column: npm . Result: 2673
Getting candidates for column: alas_2022 . Result: 10754
Getting candidates for column: DISCARDED_CNA . Result: Couldn't find column family: 'DISCARDED_CNA'
0
Getting candidates for column: suse . Result: 44
Getting candidates for column: cve5 . Result: 219391
```

There is too much difference between the PoC content and the current published one to make the right comparison. We should make sure both snapshots are taken at the same time to properly analyze the results.

**Comment by jftuduri** [comment]
Here is the snapshot generated using the VDP tools that are currently deployed in CTI, plus the POC implementation:
https://drive.google.com/file/d/1YkVmvsStvITy42SHtVSvImkkz_DzBs2b/view?usp=sharing

**Comment by pereyra-m** [comment]
## Update

I analyze the new generated snapshots.

**Comment by pereyra-m** [comment]
## Update

I've analyzed the snapshot again.

### PoC content /  4.8.0

<details><summary>Details</summary>
<p>

```
root@fb659c773250:/var/ossec#  du -h ./queue/vd/feed/ 
6.3G	./queue/vd/feed/

```

```
Getting candidates for column: nvd . Result: 342406
Getting candidates for column: suse_server_15 . Result: 261483
Getting candidates for column: suse_desktop_15 . Result: 159159
Getting candidates for column: suse_server_12 . Result: 224582
Getting candidates for column: debian . Result: 1654241
Getting candidates for column: redhat_8 . Result: 273201
Getting candidates for column: redhat_6 . Result: 166072
Getting candidates for column: redhat_7 . Result: 223647
Getting candidates for column: suse_desktop_12 . Result: 46467
Getting candidates for column: canonical . Result: 797918
Getting candidates for column: pypi . Result: 3983
Getting candidates for column: suse_server_11 . Result: 103753
Getting candidates for column: redhat_9 . Result: 248903
Getting candidates for column: alas_1 . Result: 54318
Getting candidates for column: redhat_5 . Result: 25850
Getting candidates for column: suse_desktop_11 . Result: 18847
Getting candidates for column: arch . Result: 8495
Getting candidates for column: rocky . Result: 87358
Getting candidates for column: alma_8 . Result: 78654
Getting candidates for column: alma_9 . Result: 21989
Getting candidates for column: alas_2023 . Result: 23481
Getting candidates for column: alas_2 . Result: 36792
Getting candidates for column: npm . Result: 2679
Getting candidates for column: alas_2022 . Result: 10754
Getting candidates for column: DISCARDED_CNA . Result: 3100
Getting candidates for column: suse . Result: 44
Getting candidates for column: cve5 . Result: 219005
```

</p>
</details> 


### Regular content /  4.8.0

<details><summary>Details</summary>
<p>

```
root@fb659c773250:/var/ossec# du -h ./queue/vd/feed/ 
6.0G	./queue/vd/feed/
```

```
Getting candidates for column: nvd . Result: 671076
Getting candidates for column: suse_server_15 . Result: 237797
Getting candidates for column: suse_desktop_15 . Result: 151316
Getting candidates for column: suse_server_12 . Result: 204595
Getting candidates for column: debian . Result: 1272941
Getting candidates for column: redhat_8 . Result: 187862
Getting candidates for column: redhat_6 . Result: 142236
Getting candidates for column: redhat_7 . Result: 176272
Getting candidates for column: suse_desktop_12 . Result: 46593
Getting candidates for column: canonical . Result: 592276
Getting candidates for column: pypi . Result: 3533
Getting candidates for column: suse_server_11 . Result: 104336
Getting candidates for column: redhat_9 . Result: 100095
Getting candidates for column: alas_1 . Result: 54540
Getting candidates for column: redhat_5 . Result: 26131
Getting candidates for column: suse_desktop_11 . Result: 18934
Getting candidates for column: arch . Result: 8529
Getting candidates for column: rocky . Result: 85481
Getting candidates for column: alma_8 . Result: 74638
Getting candidates for column: alma_9 . Result: 19338
Getting candidates for column: alas_2023 . Result: 21208
Getting candidates for column: alas_2 . Result: 35586
Getting candidates for column: npm . Result: 2558
Getting candidates for column: alas_2022 . Result: 10826
Getting candidates for column: DISCARDED_CNA . Result: Couldn't find column family: 'DISCARDED_CNA'
0
Getting candidates for column: suse . Result: 13
Getting candidates for column: cve5 . Result: 238656
```


</p>
</details>

**Comment by pereyra-m** [comment]
## Update

Considering that the newly generated content still contains differences hard to explain at first sight, more deep research will be performed.

**Comment by pereyra-m** [comment]
## Update

After a new analysis, the script used was fixed with a the right characters in the `grep` command. But there are considerable differences, and they are in part explained by the hardware CPEs elimination as @sebasfalcone suggested.

The key counter script gave the following results

<details><summary>Script</summary>
<p>

```python
import os

array = [
"nvd",
"suse_server_15",
"suse_desktop_15",
"suse_server_12",
"debian",
"redhat_8",
"redhat_6",
"redhat_7",
"suse_desktop_12",
"canonical",
"pypi",
"suse_server_11",
"redhat_9",
"alas_1",
"redhat_5",
"suse_desktop_11",
"arch",
"rocky",
"alma_8",
"alma_9",
"alas_2023",
"alas_2",
"npm",
"alas_2022",
"DISCARDED_CNA",
"suse"]

paths = ['/content-test/regular/vd/feed/', '/content-test/poc/vd/feed/']

for path in paths:
    print("Path: ", path)

    for column in array:
        cmd = f"""./src/build/wazuh_modules/vulnerability_scanner/testtool/rocksDBQuery/rocks_db_query_testtool -d {path} -c {column} -f /workspaces/wazuh/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs | grep ' ==> {{' | wc -l"""
        print("Getting candidates for column:", column, ". Result: ", end="", flush=True)
        os.system(cmd)

    cmd = f"""./src/build/wazuh_modules/vulnerability_scanner/testtool/rocksDBQuery/rocks_db_query_testtool -d {path} -c cve5 -f /workspaces/wazuh/src/wazuh_modules/vulnerability_scanner/schemas/cve5.fbs | grep ' ==> {{' | wc -l"""
    print("Getting candidates for column:", "cve5", ". Result: ", end="", flush=True)
    os.system(cmd)

```

</p>
</details>  

<details><summary>Results</summary>
<p>

```
Path:  /content-test/regular/vd/feed/
Getting candidates for column: nvd . Result: 671076
Getting candidates for column: suse_server_15 . Result: 237797
Getting candidates for column: suse_desktop_15 . Result: 151316
Getting candidates for column: suse_server_12 . Result: 204595
Getting candidates for column: debian . Result: 1272941
Getting candidates for column: redhat_8 . Result: 187862
Getting candidates for column: redhat_6 . Result: 142236
Getting candidates for column: redhat_7 . Result: 176272
Getting candidates for column: suse_desktop_12 . Result: 46593
Getting candidates for column: canonical . Result: 592276
Getting candidates for column: pypi . Result: 3533
Getting candidates for column: suse_server_11 . Result: 104336
Getting candidates for column: redhat_9 . Result: 100095
Getting candidates for column: alas_1 . Result: 54540
Getting candidates for column: redhat_5 . Result: 26131
Getting candidates for column: suse_desktop_11 . Result: 18934
Getting candidates for column: arch . Result: 8529
Getting candidates for column: rocky . Result: 85481
Getting candidates for column: alma_8 . Result: 74638
Getting candidates for column: alma_9 . Result: 19338
Getting candidates for column: alas_2023 . Result: 21208
Getting candidates for column: alas_2 . Result: 35586
Getting candidates for column: npm . Result: 2558
Getting candidates for column: alas_2022 . Result: 10826
Getting candidates for column: DISCARDED_CNA . Result: 0
Getting candidates for column: suse . Result: 13
Getting candidates for column: cve5 . Result: 238654
Path:  /content-test/poc/vd/feed/
Getting candidates for column: nvd . Result: 342406
Getting candidates for column: suse_server_15 . Result: 261483
Getting candidates for column: suse_desktop_15 . Result: 159159
Getting candidates for column: suse_server_12 . Result: 224582
Getting candidates for column: debian . Result: 1654241
Getting candidates for column: redhat_8 . Result: 273201
Getting candidates for column: redhat_6 . Result: 166072
Getting candidates for column: redhat_7 . Result: 223647
Getting candidates for column: suse_desktop_12 . Result: 46467
Getting candidates for column: canonical . Result: 797918
Getting candidates for column: pypi . Result: 3983
Getting candidates for column: suse_server_11 . Result: 103753
Getting candidates for column: redhat_9 . Result: 248903
Getting candidates for column: alas_1 . Result: 54318
Getting candidates for column: redhat_5 . Result: 25850
Getting candidates for column: suse_desktop_11 . Result: 18847
Getting candidates for column: arch . Result: 8495
Getting candidates for column: rocky . Result: 87358
Getting candidates for column: alma_8 . Result: 78654
Getting candidates for column: alma_9 . Result: 21989
Getting candidates for column: alas_2023 . Result: 23481
Getting candidates for column: alas_2 . Result: 36792
Getting candidates for column: npm . Result: 2679
Getting candidates for column: alas_2022 . Result: 10754
Getting candidates for column: DISCARDED_CNA . Result: 3100
Getting candidates for column: suse . Result: 44
Getting candidates for column: cve5 . Result: 218997

```

</p>
</details> 

The candidates for the ADPs are almost always more in the PoC content as expected.
But the main effort was done over `cve5` column.

A comparison script was run

<details><summary>Script</summary>
<p>

```python
import os

REGULAR_CONTENT_PATH = '/content-test/regular/vd/feed/'
POC_CONTENT_PATH = '/content-test/poc/vd/feed/'


cmd = f"""./src/build/wazuh_modules/vulnerability_scanner/testtool/rocksDBQuery/rocks_db_query_testtool -d {REGULAR_CONTENT_PATH} -c cve5 -f /workspaces/wazuh/src/wazuh_modules/vulnerability_scanner/schemas/cve5.fbs | grep ' ==> {{' """
print("Getting candidates for regular content column:", "cve5", flush=True)

# store output of the command in a variable
output = os.popen(cmd).read()
regular = output.split('\n')

print("Getting candidates for poc content column:", "cve5", flush=True)


cmd = f"""./src/build/wazuh_modules/vulnerability_scanner/testtool/rocksDBQuery/rocks_db_query_testtool -d {POC_CONTENT_PATH} -c cve5 -f /workspaces/wazuh/src/wazuh_modules/vulnerability_scanner/schemas/cve5.fbs | grep ' ==> {{' """
output = os.popen(cmd).read()
poc = output.split('\n')

poc_not_regular = 0
regular_not_poc = 0

print("Lines in POC content that are not in regular content:")
matches = []
for line in poc:
    if line not in regular:
        print(line)
        poc_not_regular += 1
    else:
        matches.append(line)

print(f"POC content has {poc_not_regular} lines that are not in regular content")

print("Lines that match in both regular and POC content:")
for line in matches:
    print(line)

print(f"There are a total of {len(matches)} matches")

print("Lines in regular content that are not in POC content:")
for line in regular:
    if line not in poc:
        print(line)
        regular_not_poc += 1

print("Regular content has", regular_not_poc, "lines that are not in POC content")

```

</p>
</details> 

And the summary is the following result

```
POC content has 3106 lines that are not in regular content

There are a total of 215892 matches

Regular content has 22763 lines that are not in POC content
```

The CVEs present in the PoC are 3100 discarded candidates and 6 rejected ones.
This was confirmed with the following script

<details><summary>Script</summary>
<p>

```python
import os

REGULAR_CONTENT_PATH = '/content-test/regular/vd/feed/'
POC_CONTENT_PATH = '/content-test/poc/vd/feed/'

DISCARDED_CNA_COUNTER = 0
TOTAL_CNA_COUNTER = 0
with open("poc_not_regular.txt") as f:
    poc = f.readlines()
    for line in poc:
        # remove whitespace characters like `\n` at the end of each line
        line = line.strip()
        cmd = f"""./src/build/wazuh_modules/vulnerability_scanner/testtool/rocksDBQuery/rocks_db_query_testtool -d {POC_CONTENT_PATH} -c cve5 -f /workspaces/wazuh/src/wazuh_modules/vulnerability_scanner/schemas/cve5.fbs -k {line} | head -n 20 """
        data = os.popen(cmd).read()
        TOTAL_CNA_COUNTER += 1
        print("Total CNA: ", TOTAL_CNA_COUNTER)
        if 'DISCARDED_CNA' in data:
            DISCARDED_CNA_COUNTER += 1
        else:
            print(data)
            print("Discarded CNA: ", DISCARDED_CNA_COUNTER)



print(f"Total CNA: {TOTAL_CNA_COUNTER}")
print(f"Discarded CNA: {DISCARDED_CNA_COUNTER}")

```

</p>
</details> 

The CVEs present in the regular content and not in the PoC are too many.
This script is proposed to check which ones are related to the hardware CPEs, but it takes too long to run

<details><summary>Script</summary>
<p>

```python
import os

REGULAR_CONTENT_PATH = '/content-test/regular/vd/feed/'
POC_CONTENT_PATH = '/content-test/poc/vd/feed/'

HARDWARE_COUNTER = 0
TOTAL_HARDWARE_COUNTER = 0
with open("regular_not_poc.txt") as f:
    poc = f.readlines()
    for line in poc:
        # remove whitespace characters like `\n` at the end of each line
        line = line.strip()
        cmd = f"""./src/build/wazuh_modules/vulnerability_scanner/testtool/rocksDBQuery/rocks_db_query_testtool -d {REGULAR_CONTENT_PATH} -c cve5 -f /workspaces/wazuh/src/wazuh_modules/vulnerability_scanner/schemas/cve5.fbs -k {line} """
        data = os.popen(cmd).read()
        TOTAL_HARDWARE_COUNTER += 1
        print("Total in progress: ", TOTAL_HARDWARE_COUNTER)
        if ':h:' in data:
            HARDWARE_COUNTER += 1
        else:
            print(data)
            print("Hardware: ", HARDWARE_COUNTER)



print(f"Total: {TOTAL_HARDWARE_COUNTER}")
print(f"Hardware: {HARDWARE_COUNTER}")

```

</p>
</details> 

The rest of the differences must be related to the missing feature of deleting entries in CTI.
Also, the published content wasn't re-built before the v4.8.0 release with the last version of the tools, making this comparison slow and complex.

**Comment by sebasfalcone** [comment]
Thanks @pereyra-m for the extensive testing!

## Conclusion
Due to the nature of the changes, it is safe to assume that the changes will not degrade the Vulnerability Scanner capabilities (it doesn't remove existing entries, it adds new ones)

The tests performed conclude that the POC content makes sense and is viable for the scanner

> [!NOTE]
> It was not easy to validate the changes because the content present in CTI contains a lot of entries that are no longer relevant. These entries are now filtered by the intelligence platform, for example, hardware vulnerabilities

CC: @Damian-Mangold
